x/programs/rust/examples/automated-market-maker/src/lib.rs

```diff
@@ -1,261 +1,249 @@
+        (token_out, token_in)
     };
 
-    context
-        .store([
-            (StateKeys::ReserveX, &reserve_x),
-            (StateKeys::ReserveY, &reserve_y),
-        ])
-        .unwrap();
+    // calculate the amount of tokens in the pool
+    let (reserve_token_in, reserve_token_out) = reserves(&token_in, &token_out);
+    assert!(reserve_token_out > 0, "insufficient liquidity");
 
-    out
-}
+    // x * y = k
+    // (x + dx) * (y - dy) = k
+    // dy = (y * dx) / (x + dx)
+    let amount_out = (reserve_token_out * amount) / (reserve_token_in + amount);
 
-fn total_supply(context: &Context<StateKeys>) -> u64 {
-    context
-        .get(StateKeys::TotalySupply)
-        .unwrap()
-        .unwrap_or_default()
-}
+    // transfer tokens fropm actor to the pool
+    // this will fail if the actor has not approved the tokens or if the actor does not have enough tokens
+    token::transfer_from(&token_in, context.actor(), *program.account(), amount);
 
-fn reserves(context: &Context<StateKeys>) -> (u64, u64) {
-    (
-        context
-            .get(StateKeys::ReserveX)
-            .unwrap()
-            .unwrap_or_default(),
-        context
-            .get(StateKeys::ReserveY)
-            .unwrap()
-            .unwrap_or_default(),
-    )
-}
-
-#[cfg(test)]
-mod tests {
-    use simulator::{Endpoint, Step, StepResponseError, TestContext};
-    use wasmlanche_sdk::ExternalCallError;
+    // transfer the amount_out to the actor
+    // we use transfer_from to update the allowance automatically
+    token::transfer_from(&token_out, *program.account(), context.actor(), amount_out);
 
-    const PROGRAM_PATH: &str = env!("PROGRAM_PATH");
+    // update the allowance for token_in
+    assert!(token::approve(
+        &token_in,
+        *program.account(),
+        reserve_token_in + amount
+    ));
 
-    #[test]
-    fn init_state() {
-        let mut simulator = simulator::ClientBuilder::new().try_build().unwrap();
+    amount_out
+}
 
-        let program_id = simulator
-            .run_step(&Step::create_program(PROGRAM_PATH))
-            .unwrap()
-            .id;
+/// Adds 'amount_x' of token_x and 'amount_y' of token_y to the pool.
+/// The ratio of the tokens must be the same as the ratio of the tokens
+/// in the pool, otherwise the function will fail.
+/// Both tokens must be approved by the actor before calling this function
+/// Returns the amount of LP shares minted
+#[public]
+pub fn add_liquidity(context: Context<StateKeys>, amount_x: u64, amount_y: u64) -> u64 {
+    let program = context.program();
+    let (token_x, token_y) = external_token_contracts(&context);
+    let lp_token = external_liquidity_token(&context);
+
+    // calculate the amount of tokens in the pool
+    let (reserve_x, reserve_y) = reserves(&token_x, &token_y);
+
+    // ensure the proper ratio
+    assert_eq!(
+        reserve_x * amount_y,
+        reserve_y * amount_x,
+        "invalid token amounts provided"
+    );
```

Comment:
Same here, if any of the multiplications overflow, then the accounting will be broken.